BUG: Fix reduce promotion with out argument

seberg opened the pull request:
This fixes promotion for reductions if `out=` is passed.  The Outputs dtype should take part in promotion here.

Due to value-based casting/promotion it does not work to pass on `out` to the legacy-promoter, because `out` may be scalar and the legacy type resolver would (de facto) ignore its precision.

So in that case, we keep using the old logic of making `out=` equivalent to an implicit `dtype=out.dtype`.

This is not quite correct in theory, because ::

    arr = np.full(5, 2**25-1, dtype=np.int64)
    np.multiply.preduce(arr, out=np.zeros((), dtype=float32))

Should naturally use float64 (as the promotion result for int64 and float32), which gives slightly different results.

This may be up for debate, but in general `out=` has normally no influence on the result of ufuncs/functions.  Reduce-like operations do feel somewhat special in this regard, but it is unclear to me that they should be.

This ensures the operation is compatible with the output argument  (i.e. minimal precision), which was intermittently broken for the  "legacy" dtype-resolution paths.

Note that we ignore the multiple/add special "upcast" paths here and that it may make sense to override this in the future if `initial=` is passed (normally `out=` should have no direct influence on the operation precision).

Closes gh-20455

seberg committed:
BUG: Fix reduce promotion with out argument

This fixes promotion for reductions if `out=` is passed.  The
Outputs dtype should take part in promotion here.

Due to value-based casting/promotion it does not work to pass on
`out` to the legacy-promoter, because `out` may be scalar and
the legacy type resolver would (de-facto) ignore its precision.

So in that case, we keep using the old logic of making `out=`
equivalent to an implicit `dtype=out.dtype`.

This is not quite correct in theory, because ::

    arr = np.full(5, 2**25-1, dtype=np.int64)
    np.multiply.preduce(arr, out=np.zeros((), dtype=float32))

Should naturally use float64 (as the promotion result for int64 and
float32), which gives slightly different results.

This may be up for debate, but in general `out=` has normally no
influence on the result of ufuncs/functions.  Reduce-like operations
do feel somewhat special in this regard, but it is unclear to me that
they should be.

This ensures the operation is compatible with the output argument
(i.e. minimal precision), which was intermittendly broken for the
"legacy" dtype-resolution paths.

Note that we ignore the multiple/add special "upcast" paths here
and that it may make sense to override this in the future if
`initial=` is passed (normally `out=` should have no direct
influence on the operation precision).

Closes gh-20455

seberg commented:
Well, value-based casting/promotion is stabbing me in the back here.  I can't use the output array as an operand, because the current type resolvers will just ignore it (depending on the value).

It seems the only solution may be to force the signature as if `dtype=` was passed :(.  I suppose conditional on no new DTypes being involved...

seberg commented:
Well, this also doesn't work for the logical reductions :(, because there `dtype=` was always just ignored, but now they warn about it in the non-reduce case.  I may need to check for this explicitly later on.

charris commented:
IIRC, the original behavior was ignore the out dtype when computing the sum, then cast the result to the `out` dtype if needed.

seberg commented:
> IIRC, the original behavior was to ignore the out dtype when computing the sum, then cast the result to the out dtype if needed.

For normal ufunc calls, yeah.  But not for reductions:
```python
In [2]: arr = np.full(5, 2**25-1, dtype=np.int64)
   ...: res_single = np.multiply.reduce(arr, out=np.zeros((), dtype=np.float32))
   ...: res_double = np.multiply.reduce(arr, out=np.zeros((), dtype=np.float64)).astype(np.float32)
   ...: res_single == res_double
Out[2]: False

In [3]: np.__version__
Out[3]: '1.21.4'
```

seberg commented:
Ah, lucky me though... The warning only kicks in if the type type signature is `(None, None, dtype)`.  It should have been `(dtype, None, dtype)` to begin with here and I just had a typo :).

So the last issue, was basically just a typo...

seberg commented:
Hmmm, maybe the best option is to just kill the legacy fallback completely for the logical ufuncs.  The new (generic) type resolution is superior anyway.

grlee77 reviewed (commented):
Thanks @seberg! The added test cases look good to me. I'm not familiar enough with the ufunc C code to comment on that part

seberg commented:
So, the thing is that I can fix this also by not forcing the signature and just using the output array as "operands" (for the legacy paths).  That should work just as well, but it will allow odder things:
```
np.logaddexp([1, 2, 3], out=np.ones((), dtype=np.uint8))
```
used to not work (because there are no integer loops), but would now work (because typical promotion currently picks a float loop instead).

I can prohibit this by ensuring that the output uses `safe` casting (which makes a lot of sense!).  If I do this, without a further special case, I will disable the following:
```
np.logaddexp([1, 2, 3], out=np.ones((), dtype=np.uint8), dtype="f8")
```
because currently, we don't check casting.

seberg commented:
I am honestly tempted to just pulling the plug on that last expression, considering that it requires both `dtype` and `out` to be passed.

seberg commented:
I am sorry this is taking so long.  It is really hard to nail everything down, because right now I think I cannot easily avoid fixing/extending functionality that I hoped to only nail down later.
The main "difficulty" is/was probably how subtly different `dtype=` is from using `out=`.

I am running into the trickier parts of NumPy's promotion/dispatching rules, for example `np.add(f4, f4, out=f8)` is different (dispatches to `f4` loop) from `np.add(f4, f4, dtype=f8)` (dispatches to `f8` loop).
So the two examples end up providing a different precision (even if that rarely matters in practice)!  And this means that it is a bit tricky to say how things should behave if an `out=` is provided.

Some things are fairly clear, if a `dtype=` is provided, I have to "honor" the "out" dtype (fix it), since this is what `dtype=` means.  If only `out=` is provided, the outputs dtype is actually irrelevant for the operation.

One thing is that I always anticipated that users could add a loop which only differs in the `out` dtype (e.g. a "fast" version for `f4 + f4 -> f8`) but now it seems that unless `dtype=` is passed, that must not be picked/preferred!  (That is, because the `f4 + f4 -> f8` loop must be the higher precision version and cannot be the lower precision one.  In theory, I suppose it would not matter too much if we pick a higher precision loop even though it was not requested, but probably not much point).

I guess that requirement/limitation above "nails it down" well enough, but I will have to sleep over it at least one more time.  There were some more confusing cases around promotion, but they seem to resolve themselves mostly.

(Aside from fixing up some of that stuff, even if it effectively should not really matter right now, one deprecation warning is currently not given and I have to see if fixing the above fixes that as well.)

seberg committed:
MAINT: Remove specialized `out=` casting loops and introduce resolvers for reductions!

seberg commented:
TL;DR: This is working now, and review would be appreciated, if it helps with review, happy to meet.  There is 1-2 things that I should at least double-check, but it should not cause large changes to the code  (maybe additions).

There are some smaller fixups that were necessary (or just seemed better).  The main parts are that:
* To fix reductions, I now always use the new promotion, because it will skip the value-based promotion stuff that is breaking things.  Because reductions used to always force a homogeneous signature, that ends up working fine.
* Since I need it for reductions, I now always register a "common dtype" default promoter.  Such a promoter should never do something wrong (but it may do something useless), so it is a good default to have in any case.  (It is not always the right one though!)
* I need a hack, to not use a new promoter for the cases where the other gives deprecation warnings still
* [ ] I need a special "object" promoter, because `np.frompyfunc` (and theoretically a user function) could have a single object-only loop, and in that case, we always use it.  (I should probably double check this one)

Two potential followups, besides general cleanup:
* That deprecation/future warning is de-facto buggy, if I can get away with skipping the deprecation, I could remove that clunk again.
* Comparisons need that deprecation/futurewarning hack, but for them the default promoter doesn't really make sense.  So I could add a specific one for them.

seberg committed:
Fix typo, fix the crucial part (yeah I forgot...) and keep a silly debug print around...

seberg commented:
OK, that was silly, I did some mistakes but mainly forgot to actually modify the part that would fix the original problem :).

On the way, I think my caching logic has some silly short-comings when promotion kicks in and I should probably fix that, but will do that in a new PR.

seberg commented:
ouch :(...  I missed, that if you do `np.logical_or.reduce(arr, out=np.empty((), dtype=arr.dtype))`, I look up a the loop with two identical input dtypes, which matches, e.g. the `d,d->?` loop, which is then not compatible with a reduction (it has to be `?,d->?` or since we do not have that directly `?,?->?` through promotion).

The way this currently works, is that we assume the output is correct (i.e. `?` here) and then search for a homogeneous loop, which ends up at the right place.  That assumption should be fine, and I think we could roll with it, although it needs to slightly special-case reductions in a way that uses the new assumptions about no "unsafe" loops (which seems totally fine).

Basically, if I end up with a loop that is not reduce-compatible, I can still steal its out-dtype and force that as the output (in `signature=`) and try again.

If we ever relax those conditions, we probably not only need to add the "which output was specified" information, but also the "was this a reduction?" information to the caching mechanism (not really super hard, but also not inviting).

seberg committed:
FIX: Add compat hack to find force a reduce-compatible loop if we found a non-reduce-compatible one

seberg commented:
OK, the last commit implements that, so hopefully that should be the last word.  I can't say I am super happy about it, but I actually think it is fine to have this "reduce" special case.  It might be nice tell about "this is a reduction" to promoters to avoid it at some point, but neither do I think it is particularly terrible.

seberg committed:
MAINT: Remove now unnecessary/bogus reduce special path

charris commented:
I'm going to let this sit for another day while you rest up :)

charris commented:
I've looked through this and don't see anything obviously wrong. Can't say much more without a lot of study and a clear idea of how it is supposed to operate. What is special about reductions that they can't just default to higher precision loops and cast the result to the output type if needed?

seberg commented:
Thanks for taking a look Chuck! Maybe ping @mhvk, in case you have time for a sanity check.

Reductions call the inner-loop effectively with:
```
np.add(res, arr, out=res)
```
so the inner-loop needs to have that pattern.  It is fine to use a higher-precision loop internally and cast the result, although now I am not sure that there will not be a bunch of back-and-forth casts going on:
```
arr = np.random.random(20)
np.add.reduce(arr, out=np.empty((), "f4")).dtype
```
works (currently and here), although it may end up with reduced precision for certain incarnations (depending on `axis=` and buffer-sizes), so I am not actually sure it should work...

That said, I think you put me on the track for another new issue:
```
np.add.reduce(arr, out=np.empty((), ">f8"))
```
Should not fail, but currently does.  Hopefully, this is a simple issue though (and likely even unrelated to the changes here).

seberg committed:
BUG: Fixup rare reducelike cast causing rejection of a valid loop

If byte-swapping occurred, this would lead to identity checks failing.
Maybe it will be easier to just not be as strict in the reduce code,
but for now fix-up the resolver to ensure identity for reducelike
operations (when possible).

Also fixes potential decref of uninitialized values.

seberg commented:
Fixed the last bug (although it was distinct from the rest as expected, also a bit annoying).

charris commented:
I'll need to think about that. What I don't see is why the dtype of an explicit out array on the python side should matter in the selection of a loop. For instance, integer sums could then have a function specific default dtype of int64 and the loop out array would then be copied into the python out array (if given) with a cast if needed. The python out array could be used for the loop if it was type identical, but sort of as a scratch array. I wonder if the default reduction dtype should be part of the ufunc?

seberg commented:
> What I don't see is why the dtype of an explicit out array on the python side should matter in the selection of a loop.

You are right of course, but it currently does.  The "logic" is that `out` has no influence, but is also the input, so it is the same as:
```
np.add(out, arr, out=out)
# which is identical in result to
np.add(out, arr)
```
(The reduction special cases kick in when the loop we pick for `np.add(out, arr)` is not compatible with a reduction.)

> I wonder if the default reduction dtype should be part of the ufunc?
Oh damn... My plan was to push this into `promotion` in the future.  Right now that doesn't matter, because we still do all the special logic ahead of dispatching.  But the current implementation is indeed not quite compatible :(.

The problem is that right now, if `np.add(out, arr)` is compatible with a reduction, I will pick it immediately, but in the case of `?,?->?` that is reduce-compatible, but not actually what we want.

Maybe I have to flag loops for "reduce-compatible?"?! Even then, I really have to signal reduction to the promoter explicitly in the future, there is no plausible way around that anymore.

seberg commented:
I lied of course:  currently reductions do not do that "default reduction dtype" thing if you provide `out=`, and they also disregard the original dtype.  But I think I can get away with changing that part.  Because, adding the original `arr` to the dispatching/promotion, will only ever pick a more precise loop.

seberg commented:
I am trying to figure out the options (right now and generally).  Right now, this works, but only because the `prod` and `multiply` is still a special cased before promotion (and user DTypes have no access to, which is something I missed!).

The solution seems to be:
* Promotion needs to know about this being a reduction, because that information cannot be passed implicitly by `descr[0] == NULL` (when out, or initial is specified).
* I am pretty sure that some loops must not be used when we have a reduction (the weird thing is that for 2 complement integers it tends to not matter for addition/multiplication).  I can think of two ways:
  1. We mark these loops as "not reduce compatible" to tell the dispatcher that it should (normally) not be used for reductions.
  2. We always force at least one promotion step when we have a reduction.

Maybe the second one is simpler?  In both cases, it requires using the reduce information in the dispatcher, and that means I need to adept the caching to not slow things down at least a bit.

The only other approach I can currently think of would be to have a special "reduce promotions", that are completely distinct.  But that seems stranger to me, and does not lessen the fact that I need to fix the cache (or dispatching could be quite noticeable for small arrays).

---

But what does that mean for right now?  I think the current state is in principle broken for cases like `prod`/`sum`, but it is at least broken in the same way that the old code was and I don't have much appetite to revert it to the old dispatcher completely either.

charris commented:
Do you plan further changes here, or is it good enough? As long as it is good enough we can regard the rest as implementation details.

> The "logic" is that out has no influence, but is also the input,

I thought the reduce loop out (and in) arrays would be new initialized arrays of `dtype=` type unrelated to the python `out` type. I suppose that didn't need to be the case, I just didn't think the python side out array played any role except to receive the output at the end.

seberg commented:
Right now, not planning any changes, mainly because I think I would end up digging into the caching and that will add more complexity here (probably not super tricky, but yet another dimension).  I don't like the current state, but I don't see "simple progress"...

>  I just didn't think the python side out array played any role except to receive the output at the end.

Yeah, this is confusing and maybe we can try to consider it legacy-behaviour and consider getting rid of it.  In <1.22, for reductions, `out=` trumps any input (I am changing to the same importance here).  Maybe we should consider ditching it, although I can live quite well with the `out` is also used as `initial` way to argue that it is OK if it has a meaning.

charris commented:
OK, let's put this in. What about your caching fixup PR, should I wait on that?

charris commented:
Thanks Sebastian.

seberg commented:
Oh, no that one for once is straight forward without any deep pits to get stuck in :).

Thanks, I may have another look over here but doubt I will be able to find anything.  Could you trigger the nightlies (I forgot how to), so we can have a sanity-check before the RC?
